crates/matrix-sdk-crypto/src/session_manager/sessions.rs

```diff
@@ -235,16 +235,19 @@ impl SessionManager {
         )
         .await?;
 
+        // A map from user ID to a map from device ID to list of algorithms.
+        let mut non_olm_devices_by_user: BTreeMap<_, BTreeMap<_, _>> = BTreeMap::new();
+
+        // A map from user id to a set of device IDs.
+        let mut bad_key_devices: BTreeMap<_, BTreeSet<_>> = BTreeMap::new();
```

Comment:
That's two new map keyed by user id that are used only for debugging purposes. Could we use a small helper struct which fields would be the three value types of these debug-purposes-only maps, instead?

```rust
struct UserDebugInfo {
  timed_out_devices: ...,
  non_olm_devices: ...,
  bad_key_devices: ...,
}

let mut user_debug_info = BTreeMap<_, UserDebugInfo>::new();
```

Reply:
Interesting idea. The difficulty is that they are a bit tied up with the decision of which log lines to write: `timed_out_devices` should not cause a warning, whereas if `non_olm_devices` or `bad_key_devices` are non-empty, we do write a warning.

Currently it's easy to check if we should write that `warn` line, because we can test `!non_olm_devices_by_user.is_empty() || !bad_key_devices.is_empty()`. If `timed_out_devices` is part of the same map, the logic for whether we write a warning is much more complicated.

So in net I don't think this gains us much; but maybe I am not entirely following your suggestion?

Reply:
I see, the `is_empty()` check is a bit annoying to do just with iterators indeed. (There might be a way using the `Peekable` iterator wrapper, and check on it if `peek().is_some()` but that's a bit contrived, even for me.)

Maybe we can then group `non_olm_devices_by_user` and `bad_key_devices` in the same struct, at least? That's still one fewer `BTreeMap`.

I'm a bit worried about all the temporary allocations here, but maybe it's not biggie; another way to confirm that would be to run the benchmark before/after, making sure the debug data structures aren't empty, to see how much of a difference that makes.

Reply:
righto. Have merged them in https://github.com/matrix-org/matrix-rust-sdk/pull/2859/commits/9b4eda86c8eb586f5d8de601e5f068c8a40fb7f6.